File: lib/core/utils/callbacks/FirebaseDatabaseCallback.java
package com.picon.utils.callbacks;

import android.app.Activity;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.picon.utils.constains.ErrorCode;
import com.picon.utils.models.KeyValue;
import com.picon.utils.providers.TaskProvider;
import com.picon.utils.response.Response;

import java.util.ArrayList;
import java.util.HashMap;

public class FirebaseDatabaseCallback extends FirebaseStorageCallback {

    public void updateRequestByKey(@NonNull DatabaseReference reference, @NonNull String key, @Nullable Object value) {
        HashMap<String, Object> map = new HashMap<>();
        map.put(key, value);
        updateRequestByKey(reference, map, response -> Log.e(TAG, "updateByKey: " + response.getMessage(), response.getException()));
    }

    public void updateRequestByKey(@NonNull DocumentReference reference, @NonNull String key, @Nullable Object value, @NonNull OnCallbackResponseListener<Void> listener) {
        HashMap<String, Object> map = new HashMap<>();
        map.put(key, value);
        updateRequestByKey(reference, map, listener);
    }

    public void updateRequestByKey(@NonNull DatabaseReference reference, @NonNull HashMap<String, Object> data, @NonNull OnCallbackResponseListener<Void> listener) {

        final Response<Void> response = new Response<>(CODE_UPDATE_BY_KEY);

        if (isInternetUnavailable()) {
            listener.onResponse(response.setErrorCode(ErrorCode.NETWORK_UNAVAILABLE));
        } else {

            reference
                    .updateChildren(data)
                    .addOnCanceledListener(mActivity, () -> listener.onResponse(response.setErrorCode(ErrorCode.CANCELED)))
                    .addOnFailureListener(mActivity, e -> listener.onResponse(response.setErrorCode(ErrorCode.FAILURE).setException(e)))
                    .addOnSuccessListener(mActivity, unused -> listener.onResponse(response.setResult(null)));
        }
    }

    public void updateRequestByKey(@NonNull DocumentReference reference, @NonNull HashMap<String, Object> data, @NonNull OnCallbackResponseListener<Void> listener) {

        final Response<Void> response = new Response<>(CODE_UPDATE_BY_KEY);

        if (isInternetUnavailable()) {
            listener.onResponse(response.setErrorCode(ErrorCode.NETWORK_UNAVAILABLE));
        } else {

            reference
                    .update(data)
                    .addOnCanceledListener(mActivity, () -> listener.onResponse(response.setErrorCode(ErrorCode.CANCELED)))
                    .addOnFailureListener(mActivity, e -> listener.onResponse(response.setErrorCode(ErrorCode.FAILURE).setException(e)))
                    .addOnSuccessListener(mActivity, unused -> listener.onResponse(response.setResult(null)));
        }
    }

    public void uploadRequestByKey(@NonNull DatabaseReference reference, @NonNull Object data, @NonNull OnCallbackResponseListener<Void> listener) {

        final Response<Void> response = new Response<>(CODE_UPLOAD_BY_KEY);

        if (isInternetUnavailable()) {
            listener.onResponse(response.setErrorCode(ErrorCode.NETWORK_UNAVAILABLE));
        } else {

            reference
                    .setValue(data)
                    .addOnCanceledListener(mActivity, () -> listener.onResponse(response.setErrorCode(ErrorCode.CANCELED)))
                    .addOnFailureListener(mActivity, e -> listener.onResponse(response.setErrorCode(ErrorCode.FAILURE).setException(e)))
                    .addOnSuccessListener(mActivity, unused -> listener.onResponse(response.setResult(null)));
        }
    }

    public void uploadRequestByKey(@NonNull DocumentReference reference, @NonNull Object data, @NonNull OnCallbackResponseListener<Void> listener) {

        final Response<Void> response = new Response<>(CODE_UPLOAD_BY_KEY);

        if (isInternetUnavailable()) {
            response.setErrorCode(ErrorCode.NETWORK_UNAVAILABLE);
            listener.onResponse(response);
        } else {

            reference
                    .set(data)
                    .addOnCanceledListener(mActivity, () -> listener.onResponse(response.setErrorCode(ErrorCode.CANCELED)))
                    .addOnFailureListener(mActivity, e -> listener.onResponse(response.setErrorCode(ErrorCode.FAILURE).setException(e)))
                    .addOnSuccessListener(mActivity, unused -> listener.onResponse(response.setResult(null)));
        }
    }

    public void uploadRequestByKey(@NonNull DatabaseReference reference, @NonNull HashMap<String, Object> data, @NonNull OnCallbackResponseListener<Void> listener) {

        final Response<Void> response = new Response<>(CODE_UPLOAD_BY_KEY);

        if (isInternetUnavailable()) {
            listener.onResponse(response.setErrorCode(ErrorCode.NETWORK_UNAVAILABLE));
        } else {

            reference
                    .setValue(data)
                    .addOnCanceledListener(mActivity, () -> listener.onResponse(response.setErrorCode(ErrorCode.CANCELED)))
                    .addOnFailureListener(mActivity, e -> listener.onResponse(response.setErrorCode(ErrorCode.FAILURE).setException(e)))
                    .addOnSuccessListener(mActivity, unused -> listener.onResponse(response.setResult(null)));
        }
    }

    public void uploadRequestByKey(@NonNull DocumentReference reference, @NonNull HashMap<String, Object> data, @NonNull OnCallbackResponseListener<Void> listener) {

        final Response<Void> response = new Response<>(CODE_UPLOAD_BY_KEY);

        if (isInternetUnavailable()) {
            response.setErrorCode(ErrorCode.NETWORK_UNAVAILABLE);
            listener.onResponse(response);
        } else {

            reference
                    .set(data)
                    .addOnCanceledListener(mActivity, () -> listener.onResponse(response.setErrorCode(ErrorCode.CANCELED)))
                    .addOnFailureListener(mActivity, e -> listener.onResponse(response.setErrorCode(ErrorCode.FAILURE).setException(e)))
                    .addOnSuccessListener(mActivity, unused -> listener.onResponse(response.setResult(null)));
        }
    }

    public void uploadRequestWithKeyValue(@NonNull DatabaseReference reference, @NonNull ArrayList<String> keys, @NonNull ArrayList<Object> values) {
        if (keys.size() > 0 && keys.size() == values.size()) {
            ArrayList<KeyValue> keyValues = new ArrayList<>();
            for (int index = 0; index < keys.size(); index++) {
                keyValues.add(new KeyValue(keys.get(0), values.get(0)));
            }
            uploadRequestWithKeyValue(reference, keyValues, response -> Log.e(TAG, "uploadRequestWithKeyValue: " + response.getMessage(), response.getException()));
        }
    }

    public void uploadRequestWithKeyValue(@NonNull CollectionReference reference, @NonNull ArrayList<String> keys, @NonNull ArrayList<Object> values) {
        if (keys.size() > 0 && keys.size() == values.size()) {
            ArrayList<KeyValue> keyValues = new ArrayList<>();
            for (int index = 0; index < keys.size(); index++) {
                keyValues.add(new KeyValue(keys.get(0), values.get(0)));
            }
            uploadRequestWithKeyValue(reference, keyValues, response -> Log.e(TAG, "uploadRequestWithKeyValue: " + response.getMessage(), response.getException()));
        }
    }

    public void uploadRequestWithKeyValue(@NonNull DatabaseReference reference, @NonNull ArrayList<KeyValue> keyValues, @NonNull OnCallbackResponseListener<Void> listener) {

        final Response<Void> response = new Response<>(CODE_UPLOAD_BY_KEY);

        if (isInternetUnavailable()) {
            listener.onResponse(response.setErrorCode(ErrorCode.NETWORK_UNAVAILABLE));
        } else {

            for (KeyValue keyValue : keyValues) {

                String key = keyValue.getKey();
                Object value = keyValue.getValue();

                reference.child(key)
                        .setValue(value)
                        .addOnCanceledListener(mActivity, () -> listener.onResponse(response.setErrorCode(ErrorCode.CANCELED)))
                        .addOnFailureListener(mActivity, e -> listener.onResponse(response.setErrorCode(ErrorCode.FAILURE).setException(e)))
                        .addOnSuccessListener(unused -> {
                            this.mTemporaryList.add(true);
                            if (TaskProvider.isComplete(keyValues.size(), mTemporaryList.size())) {
                                listener.onResponse(response.setResult(null));
                            }
                        });
            }
        }
    }

    public void uploadRequestWithKeyValue(@NonNull CollectionReference reference, @NonNull ArrayList<KeyValue> keyValues, @NonNull OnCallbackResponseListener<Void> listener) {

        final Response<Void> response = new Response<>(CODE_UPLOAD_BY_KEY);

        if (isInternetUnavailable()) {
            listener.onResponse(response.setErrorCode(ErrorCode.NETWORK_UNAVAILABLE));
        } else {

            for (KeyValue keyValue : keyValues) {

                String key = keyValue.getKey();
                Object value = keyValue.getValue();

                if (value != null) {

                    reference.document(key)
                            .set(value)
                            .addOnCanceledListener(mActivity, () -> listener.onResponse(response.setErrorCode(ErrorCode.CANCELED)))
                            .addOnFailureListener(mActivity, e -> listener.onResponse(response.setErrorCode(ErrorCode.FAILURE).setException(e)))
                            .addOnSuccessListener(unused -> {
                                this.mTemporaryList.add(true);
                                if (TaskProvider.isComplete(keyValues.size(), mTemporaryList.size())) {
                                    listener.onResponse(response.setResult(null));
                                }
                            });
                }
            }
        }
    }

    public <T> void uploadRequestWithFeedback(@NonNull DatabaseReference reference, @NonNull T data, @NonNull OnCallbackResponseListener<T> listener) {

        final Response<T> response = new Response<>(CODE_UPLOAD_BY_KEY);

        if (isInternetUnavailable()) {
            listener.onResponse(response.setErrorCode(ErrorCode.NETWORK_UNAVAILABLE));
        } else {

            reference
                    .setValue(data)
                    .addOnCanceledListener(mActivity, () -> listener.onResponse(response.setErrorCode(ErrorCode.CANCELED)))
                    .addOnFailureListener(mActivity, e -> listener.onResponse(response.setErrorCode(ErrorCode.FAILURE).setException(e)))
                    .addOnSuccessListener(mActivity, unused -> listener.onResponse(response.setResult(data)));
        }
    }

    public <T> void uploadRequestWithFeedback(@NonNull DocumentReference reference, @NonNull T data, @NonNull OnCallbackResponseListener<T> listener) {

        final Response<T> response = new Response<>(CODE_UPLOAD_BY_KEY);

        if (isInternetUnavailable()) {
            response.setErrorCode(ErrorCode.NETWORK_UNAVAILABLE);
            listener.onResponse(response);
        } else {

            reference
                    .set(data)
                    .addOnCanceledListener(mActivity, () -> listener.onResponse(response.setErrorCode(ErrorCode.CANCELED)))
                    .addOnFailureListener(mActivity, e -> listener.onResponse(response.setErrorCode(ErrorCode.FAILURE).setException(e)))
                    .addOnSuccessListener(mActivity, unused -> listener.onResponse(response.setResult(data)));
        }
    }

}

File: lib/core/utils/callbacks/firebase_database_callback.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_database/firebase_database.dart';
import 'package:flutter_utils/core/utils/enums/error_code.dart';
import 'package:flutter_utils/core/utils/providers/task_provider.dart';

import '../responses/response.dart';
import 'firebase_storage_callback.dart';

class FirebaseDatabaseCallback extends FirebaseStorageCallback {
  static const int CODE_DELETE_BY_KEY = 10010;
  static const int CODE_LOAD_BY_KEY = 10020;
  static const int CODE_UPLOAD_BY_KEY = 10030;
  static const int CODE_UPDATE_BY_KEY = 10030;
  static const int CODE_LOAD_BY_LIST = 10030;

  final List<Object> mTemporaryList = [];

  static FirebaseDatabaseCallback getInstance() {
    return FirebaseDatabaseCallback();
  }

  void deleteRequestByKey(
    dynamic reference,
    OnCallbackResponseListener<void> listener,
  ) async {
    if (reference is DocumentReference) {
      _deleteFromFS(reference, listener);
    } else if (reference is DatabaseReference) {
      _deleteFromDB(reference, listener);
    } else {
      listener.onResponse(Response().withErrorStatus(ErrorCode.INVALID));
    }
  }

  void _deleteFromDB(
    DatabaseReference reference,
    OnCallbackResponseListener<void> listener,
  ) async {
    final response = Response(CODE_DELETE_BY_KEY);

    if (await isConnected) {
      response.errorStatus = ErrorCode.NETWORK_UNAVAILABLE;
      listener.onResponse(response);
    } else {
      reference.remove().then((value) {
        listener.onResponse(response.withResult(null));
      }).onError((e, s) {
        response.errorStatus = ErrorCode.FAILURE;
        listener.onResponse(response.withException(exception: e));
      });
    }
  }

  void _deleteFromFS(
    DocumentReference reference,
    OnCallbackResponseListener<void> listener,
  ) async {
    final response = Response(CODE_DELETE_BY_KEY);

    if (await isConnected) {
      response.errorStatus = ErrorCode.NETWORK_UNAVAILABLE;
      listener.onResponse(response);
    } else {
      reference.delete().then((value) {
        listener.onResponse(response.withResult(null));
      }).onError((e, s) {
        response.errorStatus = ErrorCode.FAILURE;
        listener.onResponse(response.withException(exception: e));
      });
    }
  }

  void deleteRequestByKeys(
    dynamic reference,
    List<String> keys,
    OnCallbackResponseListener<void> listener,
  ) async {
    if (reference is CollectionReference) {
      _deletesFromFS(reference, keys, listener);
    } else if (reference is DatabaseReference) {
      _deletesFromDB(reference, keys, listener);
    } else {
      listener.onResponse(Response().withErrorStatus(ErrorCode.INVALID));
    }
  }

  void _deletesFromDB(
    DatabaseReference reference,
    List<String> keys,
    OnCallbackResponseListener<void> listener,
  ) async {
    mTemporaryList.clear();

    final response = Response(CODE_DELETE_BY_KEY);

    if (await isConnected) {
      response.errorStatus = ErrorCode.NETWORK_UNAVAILABLE;
      listener.onResponse(response);
    } else {
      for (String key in keys) {
        reference.child(key).remove().then((value) {
          mTemporaryList.add(true);
          if (TaskProvider.isComplete(keys.length, mTemporaryList.length)) {
            listener.onResponse(response.withResult(null));
          }
        }).onError((e, s) {
          response.errorStatus = ErrorCode.FAILURE;
          listener.onResponse(response.withException(exception: e));
        });
      }
    }
  }

  void _deletesFromFS(
    CollectionReference reference,
    List<String> keys,
    OnCallbackResponseListener<void> listener,
  ) async {
    mTemporaryList.clear();

    final response = Response(CODE_DELETE_BY_KEY);

    if (await isConnected) {
      response.errorStatus = ErrorCode.NETWORK_UNAVAILABLE;
      listener.onResponse(response);
    } else {
      for (String key in keys) {
        reference.doc(key).delete().then((value) {
          mTemporaryList.add(true);
          if (TaskProvider.isComplete(keys.length, mTemporaryList.length)) {
            listener.onResponse(response.withResult(null));
          }
        }).onError((e, s) {
          response.errorStatus = ErrorCode.FAILURE;
          listener.onResponse(response.withException(exception: e));
        });
      }
    }
  }

  void loadingRequestForDocument<T>(
    dynamic reference,
    T cls,
    OnCallbackResponseListener<T?> listener,
  ) async {
    if (reference is DocumentReference) {
      _loadingFromFS(reference, cls, listener);
    } else if (reference is DatabaseReference) {
      _loadingFromDB(reference, cls, listener);
    } else {
      listener.onResponse(Response<T>().withErrorStatus(ErrorCode.INVALID));
    }
  }

  void _loadingFromDB<T>(
    DatabaseReference reference,
    T cls,
    OnCallbackResponseListener<T?> listener,
  ) async {
    final response = Response<T?>(CODE_LOAD_BY_KEY);
    if (await isConnected) {
      response.errorStatus = ErrorCode.NETWORK_UNAVAILABLE;
      listener.onResponse(response);
    } else {
      listener.onResponse(response);
      reference.get().then((snapshot) {
        bool valid = snapshot.exists;
        if (valid) {
          try {
            dynamic value = snapshot.value;
            if (value != null && value.runtimeType == cls.runtimeType) {
              response.withResult(value as T);
            } else {
              response.withSnapshot(snapshot);
            }
          } on Exception catch (e) {
            response
                .withErrorStatus(ErrorCode.FAILURE)
                .withException(exception: e);
          }
        } else {
          response.withErrorStatus(ErrorCode.RESULT_NOT_FOUND);
        }
        listener.onResponse(response.withValid(valid));
      }).onError((e, s) {
        response.errorStatus = ErrorCode.FAILURE;
        listener.onResponse(response.withException(exception: e));
      });
    }
  }

  void _loadingFromFS<T>(
    DocumentReference reference,
    T cls,
    OnCallbackResponseListener<T?> listener,
  ) async {
    final response = Response<T?>(CODE_LOAD_BY_KEY);
    if (await isConnected) {
      response.errorStatus = ErrorCode.NETWORK_UNAVAILABLE;
      listener.onResponse(response);
    } else {
      listener.onResponse(response);
      reference.get().then((snapshot) {
        bool valid = snapshot.exists;
        if (valid) {
          try {
            dynamic value = snapshot.data();
            if (value != null && value.runtimeType == cls.runtimeType) {
              response.withResult(value as T);
            } else {
              response.withSnapshot(snapshot);
            }
          } on Exception catch (e) {
            response
                .withErrorStatus(ErrorCode.FAILURE)
                .withException(exception: e);
          }
        } else {
          response.withErrorStatus(ErrorCode.RESULT_NOT_FOUND);
        }
        listener.onResponse(response.withValid(valid));
      }).onError((e, s) {
        response.errorStatus = ErrorCode.FAILURE;
        listener.onResponse(response.withException(exception: e));
      });
    }
  }

  void loadingRequestForCollection<T>(
    dynamic query,
    T cls,
    OnCallbackResponseListener<List<T>> listener,
  ) async {
    if (query is CollectionReference) {
      _loadingsFromFS(query, cls, listener);
    } else if (query is DatabaseReference) {
      _loadingFromDB(query, cls, listener);
    } else {
      listener
          .onResponse(Response<List<T>>().withErrorStatus(ErrorCode.INVALID));
    }
  }

  void _loadingsFromFS<T>(
    CollectionReference query,
    T cls,
    OnCallbackResponseListener<List<T>> listener,
  ) async {
    final response = Response<List<T>>(CODE_LOAD_BY_LIST);

    if (await isConnected) {
      listener
          .onResponse(response.withErrorStatus(ErrorCode.NETWORK_UNAVAILABLE));
    } else {
      listener.onResponse(response);
      query.get().then((snapshots) {
        bool valid = snapshots.size > 0;
        if (valid) {
          List<T> list = [];
          for (DocumentSnapshot snapshot in snapshots.docs) {
            if (snapshot.exists) {
              try {
                dynamic value = snapshot.data();
                if (value != null && value.runtimeType == cls.runtimeType) {
                  list.add(value as T);
                }
              } on Exception catch (e) {
                response
                    .withErrorStatus(ErrorCode.FAILURE)
                    .withException(exception: e);
              }
            }
          }
        } else {
          response.withErrorStatus(ErrorCode.RESULT_NOT_FOUND);
        }
        listener.onResponse(response.withValid(valid));
      }).onError((e, s) {
        response.errorStatus = ErrorCode.FAILURE;
        listener.onResponse(response.withException(exception: e));
      });
    }
  }

  void _loadingsFromDB<T>(
    DatabaseReference query,
    T cls,
    OnCallbackResponseListener<List<T>> listener,
  ) async {
    final response = Response<List<T>>(CODE_LOAD_BY_LIST);

    if (await isConnected) {
      listener
          .onResponse(response.withErrorStatus(ErrorCode.NETWORK_UNAVAILABLE));
    } else {
      listener.onResponse(response);
      query.get().then((snapshots) {
        bool valid = snapshots.children.isNotEmpty;
        if (valid) {
          List<T> list = [];
          for (DataSnapshot snapshot in snapshots.children) {
            if (snapshot.exists) {
              try {
                dynamic value = snapshot.value;
                if (value != null && value.runtimeType == cls.runtimeType) {
                  list.add(value as T);
                }
              } on Exception catch (e) {
                response
                    .withErrorStatus(ErrorCode.FAILURE)
                    .withException(exception: e);
              }
            }
          }
        } else {
          response.withErrorStatus(ErrorCode.RESULT_NOT_FOUND);
        }
        listener.onResponse(response.withValid(valid));
      }).onError((e, s) {
        response.errorStatus = ErrorCode.FAILURE;
        listener.onResponse(response.withException(exception: e));
      });
    }
  }
}
